### PR TITLE
API Processing Optimizations

### LINKED ISSUE #480: [Data Complexity] Phase 1 Follow-up: Use optimized format consistently across client-side
**Is your feature request related to a problem? Please describe.**
As part of our work to increase app scalability in the face of increasing data complexity and volume, we switched the backend to using a new optimized metric file format provided by the data platform. It is working well in production so far!

However, as some in-progress PRs can attest, and as our design document predicted would eventually happen, we are seeing that new features adding on still more dimensionality to certain metrics are causing the client to bog down and even fail to load on machines with lower memory. This is because the frontend is taking the optimized format and immediately expanding back into the previous, wasteful format. This was done simply to make the change simpler and faster, but we should now go a step further and change the frontend to filter on and present from the optimized format itself.

**Describe the solution you'd like**
Update the client side code in `src/api/metrics`, and calling code in chart components, to work directly with the optimized format. This should put a nice damper on memory consumption on the client and because the optimized format is just arrays and arithmetic, filtering data points should seemingly be even faster because we'll be able to perform constant time lookups instead of linear scans with each new filter interaction.

We should never expand the optimized format into the non-optimized format, but the one chart that still uses the non-optimized format (the Revocation Analysis Caseload Table) should continue unchanged.

**Describe alternatives you've considered**
N/A. This is the alternative to _not_ doing this work in #464.

**Additional context**
See the original data complexity design doc for more information on the implementation.

### PR DESCRIPTION
**Pull request by jessex**
## Description of the change

This performs optimization on the client-side by ensuring that the optimized metric file format is not eagerly expanded into the comparatively "verbose" array of objects, instead having chart components hold the optimized format in memory, filtering against that format, and expanding into an array of objects only for the filtered subset.

This also includes some additional optimizations, including only "unflattening" the compact value matrix up-front upon receiving the API response instead of "unflattening" each time a filter operation takes place.

Another optimization is to change the UX of the Lantern view: instead of proactively loading all "below-the-fold" revocation charts proactively and then hiding all but the selected one (which was originally done to ensure that switching between charts was immediate but means that we hold all of those API responses in memory and perform filtering for the hidden charts), we now only load a "below-the-fold" chart when it's selected, which cuts filter time and memory consumption roughly in half and scales better as we add more charts down in that section of the view.

One important note is that we have only applied the top-level change for the Lantern view: the non-Lantern views are still eagerly expanding, because the memory requirements are significantly smaller and it provides a better user experience. It's a reasonable debate, however, whether a _slightly_ more noticeable latency in filtering could be traded for more consistent flow and simpler code.

**Note: This PR is best navigated commit-by-commit.**

**TODO**:
- [x] Improve test coverage, particularly around `dataFilters.js` and `optimizedMetricFileParserHelpers.js`
- [x] Current testing with live data suggests that these changes are not yet actually required in staging or production for either Lantern state (though more testing should be done in the next day or so). However, the #477 PR includes drastically expanded fixture files which _do_ require these changes to not ground demo mode to a halt. We intend to truncate the fixture files in that PR before landing it, and moving the expanded fixture files into this PR as a proof of concept.

**Notes on future steps**:
- With these changes, there is now a more noticeable latency in the filter operation, as the core trade-off was less memory consumed for more time in filtering. So continuing to make filtering as quick as possible is paramount.
  - In addition to that, it would be useful to refactor the code around API processing a bit to permit showing a loading spinner while _filtering_ is happening.
- Profiling shows that the logic to unflatten the compact value matrix is fairly expensive, so while we are now only performing it once per API request instead of once per filter operation, it still adds some non-zero amount of observed latency on page load.
- The actual filtering code against the optimized format in `dataFilters.js` can be made more efficient, probably by adapting a different data structure for the in-memory representation of the optimized format (as opposed to the naive translation into a multi-dimensional array). Testing different approaches to iteration did not appear to yield much benefit, so a different data structure is probably needed.
- There is still some code redundancy between `dataFilters.js` and `optimizedMetricFileParser.js` that could be DRY'd up a bit.
- Ultimately, there is still a fundamental scale issue in that increased numbers of data points leads to more memory consumed (though less after this change) and more time required to perform filtering (more after this change). That fundamental issue cannot be resolved without delving into Phase 2 of the original data complexity design doc.

## Type of change

- [ ] Bug fix (non-breaking change that fixes an issue)
- [ ] New feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] Configuration change (adjusts configuration to achieve some end related to functionality, development, performance, or security)

## Related issues

Closes #480 

### Development

These boxes should be checked by the submitter prior to merging:

- [x] Manual testing against realistic data has been performed locally

### Code review

These boxes should be checked by reviewers prior to merging:

- [x] This pull request has a descriptive title and information useful to a reviewer
- [x] This pull request has been moved out of a Draft state, has no "Work In Progress" label, and has assigned reviewers
- [x] Potential security implications or infrastructural changes have been considered, if relevant

### CODE REVIEW

**Review by jovergaag — COMMENTED** [review]
Wow, there's some great work and some gnarly multi-dimensional array work in here! I have lots of comments about non critical things which can be easily addressed.

My initial reaction here is that we need to test some of these complex functions. I have spent the better part of the afternoon walking through them to understand exactly what is going on, and am still not sure I am confident that they are working as expected. I know that there is some time pressure to get this in for the PA launch (at least thats what I think was mentioned on Monday, or was it because demo data was not loading with Vlad's new work?), so lets talk about the priority of this performance enhancement vs potentially introducing bugs that will be very hard to spot and debug later down the line. 

Thoughts? I also come from a place where EVERYTHING was tested, and if that's just not where we are with this app yet, then I can jump on board with it.

Fully on board with the reorganization and loading the charts below the fold once they're selected, great optimization there!

**Comment by jessex** [comment]
Thanks for the review, @jovergaag! Quick status update and context:
- We no longer believe this is launch-blocking so fortunately there is not an urgent need to rush this in
- Definitely this PR needs more test coverage and I neglected to list that in the outstanding TODOs in the description. Just updated that
- The testing I did here was side-by-side visual assessments: run the app before and after the changes and test a variety of toggle states to ensure that the visualized data is identical, which it has been in the testing I've done, fortunately
- Our test coverage on this app is woeful in comparison to everything else we've built, which is a product of its Frankenstein-esque heritage. Improving test coverage in the app is definitely a priority

File: src/components/charts/new_revocations/RevocationsByRiskLevel/RevocationsByRiskLevel.js [hunk 1]
```diff
@@ -74,20 +75,24 @@ const RevocationsByRiskLevel = ({
...
-    treatCategoryAllAsAbsent
-  );
+  const filterFn = dataFilter(skippedFilters, treatCategoryAllAsAbsent);
 
   const riskLevelCounts = pipe(
+    (metricFile) =>
+      filterOptimizedDataFormat(
+        metricFile.flattenedValueMatrix,
+        metricFile.metadata,
+        {},
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/components/charts/new_revocations/RevocationsByRiskLevel/RevocationsByRiskLevel.js]: Outdated: this is from a previous commit.

**Comment by jovergaag** [comment]
@jessex Great, good to know there is not a need to rush this.

I know you have a lot on your plate. Let me know if you would like me to write some tests for this and/or try out different data structures for the in memory data, or if theres another way I can help get it over the finish line if you need to focus elsewhere.

File: src/api/metrics/optimizedFormatHelpers.js [hunk 2]
```diff
@@ -75,6 +75,17 @@ const unflattenValues = (flattenedValues, totalDataPoints) => {
   return unflattened;
 };
 
+/**
+ * Convers the given optimized array as a singular string into an unflattened matrix.
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/api/metrics/optimizedFormatHelpers.js:79]: Thanks!

File: src/components/charts/new_revocations/helpers.js [hunk 3]
```diff
@@ -21,96 +21,119 @@ import {
...
-  return (
-    filters.filter((value) => value.toLowerCase() === field.toLowerCase())
-      .length !== 0
-  );
-};
+export const matchesTopLevelFilters = (filters) => (
+  skippedFilters = [],
+  treatCategoryAllAsAbsent = false
+) => (item, dimensionKey = undefined) => {
+  if (
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/components/charts/new_revocations/helpers.js:35]: I'm not sure I follow. The current logic does not guarantee returning true or false if `dimensionKey` is undefined. If `dimensionKey === undefined` then the first part of each distinct block evaluates to true, but we still depend on the rest. If `dimensionKey !== undefined`, then the first part of the check can still evaluate to true if the dimensionKey matches one of the values. Am I misunderstanding?

File: src/api/metrics/optimizedFormatHelpers.js [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+/**
+ * Validates the given metric file metadata to ensure that certain assumptions are met:
+ *   - That metadata.total_data_points is defined and is numeric
+ *   - That metadata.value_keys is defined and is a non-empty array
+ *   - That metadata.dimension_manifest is defined and is a non-empty array
+ *   - That each entry in metadata.dimension_manifest is itself an array of length 2 (dimension_key, [possible_values])
+ *   - That each array of possible values in metadata.dimension_manifest is a non-empty array
+ */
+const validateMetadata = (metadata) => {
+  if (metadata.total_data_points === undefined) {
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/api/metrics/optimizedFormatHelpers.js:98]: It's theoretically possible, if the data platform failed in such a way that it unexpectedly wrote a `total_data_points` value of "null" and the built-in json deserialization in JS turned it into `null` over here. So to be safe, I'm going to expand this check to be `metadata.total_data_points === null` which will check undefined, as well.

File: src/api/metrics/optimizedFormatHelpers.js [hunk 5]
```diff
@@ -0,0 +1,148 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/api/metrics/optimizedFormatHelpers.js:17]: Done!

File: src/utils/charts/dataFilters.js [hunk 6]
```diff
@@ -0,0 +1,106 @@
...
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+// =============================================================================
+
```
[the remark this answers is not in the document]
Comment (jessex) [reply src/utils/charts/dataFilters.js:17]: Done!

**Comment by jessex** [comment]
@jovergaag: thanks for the review! I've gone through all of your feedback, increased test coverage, and rebased against master. See the latest commits.

Another thing for you to know in case you want to test: check out that unchecked TODO in my PR description about adding in a larger set of fixture files that make this a more interesting PR by increasing the amount of memory used in demo mode. If you want to test that out, you can fetch [this branch](https://github.com/Recidiviz/pulse-dashboard/tree/jessex/updated-revocation-matrix-fixtures-sep-2020) and cherry-pick the head commit (the one that updates all of the Lantern fixture files) onto my branch (or a branch from my branch, however you prefer), and then run the app in demo mode.

File: src/components/charts/new_revocations/helpers.js (hunk 3 above)
[the remark this answers is not in the document]
Comment (jovergaag) [reply src/components/charts/new_revocations/helpers.js:35]: You're right, I was mis-reading it.

**Review by jovergaag — APPROVED** [review]
@jessex Bravo! Thanks for taking the time to write those tests. 

The performance improvement is very impressive with those expanded fixture files. My browser is not at all handling those new fixtures without the optimizations. I agree that a loading indicator would be helpful for the filters for a better user experience, I'll open an issue for that to make sure we don't lose track of it.

**Comment by jessex** [comment]
@jovergaag and @sychang, from your own testing and consideration, would you prefer to include this in the launch for PA, or defer it to post-launch? I could see arguments in either direction (less chance of regression risk to wait until after launch, and it's not strictly necessary for launch, but it does help avoid unexpected page halts in the event that we have an unexpected jump in data volume in the near future). If you want it to go into launch, I'll go ahead and merge it now.

**Comment by sychang** [comment]
Y'all will have a better sense of the technical risks than me so happy to defer if the risks feel substantial enough to hold off. Based on the information I have, I would go ahead and merge so we'll have it in place to support the highest priority fast follows that are going to come soon after. Thank you both!

**Comment by jovergaag** [comment]
@jessex Sorry for the delayed response. I would prefer to wait until after PA launch. There are several things lined up that require my attention before launch, so if this were to land and we ended up finding a data bug in the filters or charts, I will be pretty stretched to address all the things. Another downside, as you can see, is all the merge conflicts building up, but to me it still seems more wise to wait.

**Comment by jessex** [comment]
I agree, @jovergaag, but wanted the Lantern team to make that call themselves. Thanks for chiming in! I'll make sure the conflicts are resolved this week so you have the ability to move more quickly on it post-launch.

**Comment by sychang** [comment]
Circling back to second the decision to wait since now the admission type filter is clearly out of scope and there are several small changes we are trying to land before launch on Jen's plate!

**Comment by jovergaag** [comment]
@jessex Update on this. After merging master yesterday I was doing some performance analysis and more QA and ended up finding a few bugs. Two were easy-ish to fix - but another bug in the filtering ended up being very difficult to get to the root of  because of the added complexity of the optimized format data being passed through the filters. This is added layer of complexity over the filtering logic was the primary reason I suggested not to merge this before PA launch - and I'm even more nervous about it now given my difficulty debugging yesterday. It turns out it was a bug in the `metadata.dimension_manifest` of one of the metric files.  `filterOptimizedDataFormat` is using the `dimension_manifest` and `value_keys` differently than `expandMetricRepresentation` so the expanded data vs unflattenened data had different filtering results. 

Given all this, I'm inclined to again suggest holding off on merging this to see if we can improve IE11 performance by optimizing the data itself. I would be interested in your thoughts on all of this since you have much more history with this PR and the data/filters themselves.

NOTE: Do not merge this as is - this code does not have a bug but the metric file dimension manifest is still incorrect so merging this is dependent on fixing that file, and we would need to QA this again.

**Comment by jessex** [comment]
>Given all this, I'm inclined to again suggest holding off on merging this to see if we can improve IE11 performance by optimizing the data itself.

I think I'd need to know what you have in mind in terms of "optimizing the data." But I think I'd say that if we have production-facing IE11 performance issues and we are confident that this PR alleviates them but the challenge is addressing the complexity that this certainly adds, then we should invest in managing that complexity instead of deferring this PR for long. If you have other small wins you can get for IE11 then I'm on board with knocking out low-hanging fruit! I just wouldn't delay this for longer than feels reasonable. I think the optimizations in this PR are certainly good things and we will want them going forward, but we should feel comfortable with how they are implemented.

>It turns out it was a bug in the metadata.dimension_manifest of one of the metric files. filterOptimizedDataFormat is using the dimension_manifest and value_keys differently than expandMetricRepresentation so the expanded data vs unflattenened data had different filtering results.

Are you referring to fixture files or metric files actually produced by the calculation pipeline? If the former, then I stick with what I said above about just managing the complexity and coming to feel good about the state of the code. If the latter, then that's potentially a different situation entirely.

**Comment by jovergaag** [comment]
@jessex The bug is in the metadata of the file (`revocations_matrix_distribution_by_risk_level`) produced by the calculation pipeline. The same bug is in both the staging and production metric file. Demo data as it exists today is correct so we would not have caught this bug if using only demo data.

[Here](revocations_matrix_distribution_by_risk_level) is the task that @ageiduschek and @jjdressel11 and I discussed to optimize the API data. I'm hoping to start a spike to investigate its viability today or tomorrow.

**Comment by jessex** [comment]
Heads up: I think that task link from your previous comment is malformed, @jovergaag.

**Comment by jovergaag** [comment]
@jessex The link is fixed.

**Comment by jessex** [comment]
Thanks, @jovergaag! I certainly leave it to Team Lantern to decide how to prioritize the two efforts. At first glance, I am _imagining_ that the investigation required in that linked ticket might be more intensive than the investigation involved in rectifying whatever discrepancy exists in the metadata for the `revocations_matrix_distribution_by_risk_level` file (is there a ticket for that, too, or is it bundled as part of this PR?) but I could be wrong, and I trust your instincts.

Let me know if and when you want any help/testing/review on either front.

**Comment by jovergaag** [comment]
@jessex Once the metrics are exported based on today's prod deploy, and after a good thorough QA with those files, this will be ready to go. I approved the PR already, but if you want to take a look at the commits I added after the approval I would appreciate it.

**Review by jessex — COMMENTED** [review]
You've got my approval, @jovergaag! Thanks so much for taking this over the finishline. Merge when ready.

**Pull request merged**